lxd/container_lxc.go

```diff
@@ -3224,57 +3220,83 @@ func (c *containerLXC) Update(args containerArgs, userRequested bool) error {
+			oldProfileRootDiskDevices = append(oldProfileRootDiskDevices, k)
 		}
+	}
 
-		if m["size"] != oldRootfsSize {
-			size, err := shared.ParseByteSizeString(m["size"])
-			if err != nil {
-				return err
-			}
-
-			err = c.storage.ContainerSetQuota(c, size)
-			if err != nil {
-				return err
-			}
+	// Retrieve new root disk devices.
+	newLocalRootDiskDeviceKey, newLocalRootDiskDevice := containerGetRootDiskDevice(c.localDevices)
+	var newProfileRootDiskDevices []string
+	for k, v := range c.expandedDevices {
+		if isRootDiskDevice(v) && k != newLocalRootDiskDeviceKey && !shared.StringInSlice(k, newProfileRootDiskDevices) {
+			newProfileRootDiskDevices = append(newProfileRootDiskDevices, k)
 		}
 	}
 
-	// Confirm that the storage pool didn't change.
-	var oldRootfs types.Device
-	for _, m := range oldExpandedDevices {
-		if m["type"] == "disk" && m["path"] == "/" {
-			oldRootfs = m
-			break
+	// Verify root disk devices. (Be specific with error messages.)
+	var oldRootDiskDeviceKey string
+	var newRootDiskDeviceKey string
+	if oldLocalRootDiskDevice["pool"] != "" {
+		oldRootDiskDeviceKey = oldLocalRootDiskDeviceKey
+		newRootDiskDeviceKey = newLocalRootDiskDeviceKey
+
+		if newLocalRootDiskDevice["pool"] == "" {
+			if len(newProfileRootDiskDevices) == 0 {
+				return fmt.Errorf("Update will cause the container to rely on a profile's root disk device but none was found.")
+			} else if len(newProfileRootDiskDevices) > 1 {
+				return fmt.Errorf("Update will cause the container to rely on a profile's root disk device but conflicting devices were found.")
+			} else if c.expandedDevices[newProfileRootDiskDevices[0]]["pool"] != oldLocalRootDiskDevice["pool"] {
+				newRootDiskDeviceKey = newProfileRootDiskDevices[0]
+				return fmt.Errorf("Using the profile's root disk device would change the storage pool of the container.")
+			}
+		}
+	} else {
+		// This branch should allow us to cover cases where a container
+		// didn't have root disk device before for whatever reason. As
+		// long as there is a root disk device in one of the local or
+		// profile devices we're good.
+		if newLocalRootDiskDevice["pool"] != "" {
+			newRootDiskDeviceKey = newLocalRootDiskDeviceKey
+
+			if len(oldProfileRootDiskDevices) > 0 {
+				oldRootDiskDeviceKey = oldProfileRootDiskDevices[0]
+				if oldExpandedDevices[oldRootDiskDeviceKey]["pool"] != newLocalRootDiskDevice["pool"] {
+					return fmt.Errorf("The new local root disk device would change the storage pool of the container.")
+				}
+			}
+		} else {
+			if len(newProfileRootDiskDevices) == 0 {
+				return fmt.Errorf("Update will cause the container to rely on a profile's root disk device but none was found.")
+			} else if len(newProfileRootDiskDevices) > 1 {
+				return fmt.Errorf("Using the profile's root disk device would change the storage pool of the container.")
+			}
+			newRootDiskDeviceKey = newProfileRootDiskDevices[0]
 		}
 	}
```

Comment:
Just a note that I need to review this whole chunk of code. The errors messages all look good but the new/old logic is pretty complex and I need to read through it carefully.